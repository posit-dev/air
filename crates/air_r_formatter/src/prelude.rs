pub(crate) use crate::{AsFormat as _, FormatNodeRule, RFormatContext, RFormatter};
pub use biome_formatter::prelude::*;
pub use biome_rowan::{AstNode as _, AstNodeList as _, AstSeparatedList as _};
