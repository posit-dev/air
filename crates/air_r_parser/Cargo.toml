[package]
authors.workspace = true
categories.workspace = true
description = "Air R parser"
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
name = "air_r_parser"
publish = false
repository.workspace = true
rust-version.workspace = true
version = "0.0.0"

[dependencies]
air_r_factory = { workspace = true }
air_r_syntax = { workspace = true }
biome_diagnostics = { workspace = true }
biome_parser = { workspace = true }
biome_rowan = { workspace = true }
biome_unicode_table = { workspace = true }
serde = { workspace = true, features = ["derive"] }
tracing = { workspace = true }
tree-sitter = { workspace = true }
tree-sitter-r = { workspace = true }

[dev-dependencies]
biome_console = { workspace = true }
biome_diagnostics = { workspace = true }
insta = { workspace = true }
line_ending = { workspace = true }
tests_macros = { workspace = true }

# cargo-workspaces metadata
[package.metadata.workspaces]
independent = true

[lints]
workspace = true
