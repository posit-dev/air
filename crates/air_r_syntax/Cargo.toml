[package]
name = "air_r_syntax"
version = "0.0.0"
authors.workspace = true
edition.workspace = true
rust-version.workspace = true
description = "SyntaxKind and common rowan definitions for air_r_parser"
homepage.workspace = true
repository.workspace = true
license.workspace = true
keywords.workspace = true
categories.workspace = true
publish = false

[dependencies]
biome_rowan = { workspace = true, features = ["serde"] }
serde = { workspace = true, features = ["derive"] }

[lints]
workspace = true
