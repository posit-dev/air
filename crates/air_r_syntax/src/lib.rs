#[macro_use]
mod generated;
mod file_source;
pub mod string_ext;
mod syntax_node;

pub use self::generated::*;
pub use biome_rowan::{TextLen, TextRange, TextSize, TokenAtOffset, TriviaPieceKind, WalkEvent};
pub use file_source::RFileSource;
pub use syntax_node::*;

use biome_rowan::{RawSyntaxKind, SyntaxKind, TokenText};

impl From<u16> for RSyntaxKind {
    fn from(d: u16) -> RSyntaxKind {
        assert!(d <= (RSyntaxKind::__LAST as u16));
        unsafe { std::mem::transmute::<u16, RSyntaxKind>(d) }
    }
}

impl From<RSyntaxKind> for u16 {
    fn from(k: RSyntaxKind) -> u16 {
        k as u16
    }
}

impl biome_rowan::SyntaxKind for RSyntaxKind {
    const TOMBSTONE: Self = RSyntaxKind::TOMBSTONE;
    const EOF: Self = RSyntaxKind::EOF;

    fn is_bogus(&self) -> bool {
        match self {
            RSyntaxKind::R_BOGUS
            | RSyntaxKind::R_BOGUS_VALUE
            | RSyntaxKind::R_BOGUS_EXPRESSION
            | RSyntaxKind::R_BOGUS_PARAMETER
            | RSyntaxKind::R_BOGUS_ARGUMENT => true,

            RSyntaxKind::TOMBSTONE
            | RSyntaxKind::EOF
            | RSyntaxKind::UNICODE_BOM
            | RSyntaxKind::SEMICOLON
            | RSyntaxKind::COMMA
            | RSyntaxKind::L_CURLY
            | RSyntaxKind::R_CURLY
            | RSyntaxKind::L_BRACK
            | RSyntaxKind::R_BRACK
            | RSyntaxKind::L_PAREN
            | RSyntaxKind::R_PAREN
            | RSyntaxKind::PLUS
            | RSyntaxKind::EQUAL
            | RSyntaxKind::DOTS
            | RSyntaxKind::FUNCTION_KW
            | RSyntaxKind::FOR_KW
            | RSyntaxKind::IN_KW
            | RSyntaxKind::WHILE_KW
            | RSyntaxKind::REPEAT_KW
            | RSyntaxKind::IF_KW
            | RSyntaxKind::ELSE_KW
            | RSyntaxKind::R_INTEGER_LITERAL
            | RSyntaxKind::R_DOUBLE_LITERAL
            | RSyntaxKind::R_COMPLEX_LITERAL
            | RSyntaxKind::R_STRING_LITERAL
            | RSyntaxKind::R_LOGICAL_LITERAL
            | RSyntaxKind::R_NULL_LITERAL
            | RSyntaxKind::NEWLINE
            | RSyntaxKind::WHITESPACE
            | RSyntaxKind::IDENT
            | RSyntaxKind::COMMENT
            | RSyntaxKind::BACKSLASH
            | RSyntaxKind::R_ROOT
            | RSyntaxKind::R_DOTS
            | RSyntaxKind::R_IDENTIFIER
            | RSyntaxKind::R_BINARY_EXPRESSION
            | RSyntaxKind::R_FUNCTION_DEFINITION
            | RSyntaxKind::R_PARAMETERS
            | RSyntaxKind::R_PARAMETER_LIST
            | RSyntaxKind::R_DOTS_PARAMETER
            | RSyntaxKind::R_IDENTIFIER_PARAMETER
            | RSyntaxKind::R_DEFAULT_PARAMETER
            | RSyntaxKind::R_IF_STATEMENT
            | RSyntaxKind::R_ELSE_CLAUSE
            | RSyntaxKind::R_FOR_STATEMENT
            | RSyntaxKind::R_WHILE_STATEMENT
            | RSyntaxKind::R_REPEAT_STATEMENT
            | RSyntaxKind::R_BRACED_EXPRESSIONS
            | RSyntaxKind::R_PARENTHESIZED_EXPRESSION
            | RSyntaxKind::R_CALL
            | RSyntaxKind::R_CALL_ARGUMENTS
            | RSyntaxKind::R_ARGUMENT_LIST
            | RSyntaxKind::R_NAMED_ARGUMENT
            | RSyntaxKind::R_UNNAMED_ARGUMENT
            | RSyntaxKind::R_DOTS_ARGUMENT
            | RSyntaxKind::R_HOLE_ARGUMENT
            | RSyntaxKind::R_EXPRESSION_LIST
            | RSyntaxKind::R_INTEGER_VALUE
            | RSyntaxKind::R_DOUBLE_VALUE
            | RSyntaxKind::R_COMPLEX_VALUE
            | RSyntaxKind::R_STRING_VALUE
            | RSyntaxKind::R_LOGICAL_VALUE
            | RSyntaxKind::R_NULL_VALUE
            | RSyntaxKind::__LAST => false,
        }
    }

    fn to_bogus(&self) -> Self {
        match self {
            // Bogus value
            RSyntaxKind::R_INTEGER_VALUE => RSyntaxKind::R_BOGUS_VALUE,
            RSyntaxKind::R_DOUBLE_VALUE => RSyntaxKind::R_BOGUS_VALUE,
            RSyntaxKind::R_COMPLEX_VALUE => RSyntaxKind::R_BOGUS_VALUE,
            RSyntaxKind::R_STRING_VALUE => RSyntaxKind::R_BOGUS_VALUE,
            RSyntaxKind::R_LOGICAL_VALUE => RSyntaxKind::R_BOGUS_VALUE,
            RSyntaxKind::R_NULL_VALUE => RSyntaxKind::R_BOGUS_VALUE,
            RSyntaxKind::R_BOGUS_VALUE => RSyntaxKind::R_BOGUS_VALUE,

            // Bogus parameter
            RSyntaxKind::R_DOTS_PARAMETER => RSyntaxKind::R_BOGUS_PARAMETER,
            RSyntaxKind::R_IDENTIFIER_PARAMETER => RSyntaxKind::R_BOGUS_PARAMETER,
            RSyntaxKind::R_DEFAULT_PARAMETER => RSyntaxKind::R_BOGUS_PARAMETER,
            RSyntaxKind::R_BOGUS_PARAMETER => RSyntaxKind::R_BOGUS_PARAMETER,

            // Bogus expression
            RSyntaxKind::R_BINARY_EXPRESSION => RSyntaxKind::R_BOGUS_EXPRESSION,
            RSyntaxKind::R_BOGUS_EXPRESSION => RSyntaxKind::R_BOGUS_EXPRESSION,

            // Bogus argument
            RSyntaxKind::R_NAMED_ARGUMENT => RSyntaxKind::R_BOGUS_ARGUMENT,
            RSyntaxKind::R_UNNAMED_ARGUMENT => RSyntaxKind::R_BOGUS_ARGUMENT,
            RSyntaxKind::R_DOTS_ARGUMENT => RSyntaxKind::R_BOGUS_ARGUMENT,
            RSyntaxKind::R_HOLE_ARGUMENT => RSyntaxKind::R_BOGUS_ARGUMENT,
            RSyntaxKind::R_BOGUS_ARGUMENT => RSyntaxKind::R_BOGUS_ARGUMENT,

            // Bogus
            RSyntaxKind::TOMBSTONE => RSyntaxKind::R_BOGUS,
            RSyntaxKind::EOF => RSyntaxKind::R_BOGUS,
            RSyntaxKind::UNICODE_BOM => RSyntaxKind::R_BOGUS,
            RSyntaxKind::SEMICOLON => RSyntaxKind::R_BOGUS,
            RSyntaxKind::COMMA => RSyntaxKind::R_BOGUS,
            RSyntaxKind::L_CURLY => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_CURLY => RSyntaxKind::R_BOGUS,
            RSyntaxKind::L_BRACK => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_BRACK => RSyntaxKind::R_BOGUS,
            RSyntaxKind::L_PAREN => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_PAREN => RSyntaxKind::R_BOGUS,
            RSyntaxKind::PLUS => RSyntaxKind::R_BOGUS,
            RSyntaxKind::EQUAL => RSyntaxKind::R_BOGUS,
            RSyntaxKind::DOTS => RSyntaxKind::R_BOGUS,
            RSyntaxKind::FUNCTION_KW => RSyntaxKind::R_BOGUS,
            RSyntaxKind::FOR_KW => RSyntaxKind::R_BOGUS,
            RSyntaxKind::IN_KW => RSyntaxKind::R_BOGUS,
            RSyntaxKind::WHILE_KW => RSyntaxKind::R_BOGUS,
            RSyntaxKind::REPEAT_KW => RSyntaxKind::R_BOGUS,
            RSyntaxKind::IF_KW => RSyntaxKind::R_BOGUS,
            RSyntaxKind::ELSE_KW => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_INTEGER_LITERAL => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_DOUBLE_LITERAL => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_COMPLEX_LITERAL => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_STRING_LITERAL => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_LOGICAL_LITERAL => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_NULL_LITERAL => RSyntaxKind::R_BOGUS,
            RSyntaxKind::NEWLINE => RSyntaxKind::R_BOGUS,
            RSyntaxKind::WHITESPACE => RSyntaxKind::R_BOGUS,
            RSyntaxKind::IDENT => RSyntaxKind::R_BOGUS,
            RSyntaxKind::COMMENT => RSyntaxKind::R_BOGUS,
            RSyntaxKind::BACKSLASH => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_ROOT => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_IDENTIFIER => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_DOTS => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_FUNCTION_DEFINITION => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_PARAMETERS => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_PARAMETER_LIST => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_IF_STATEMENT => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_ELSE_CLAUSE => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_FOR_STATEMENT => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_WHILE_STATEMENT => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_REPEAT_STATEMENT => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_BRACED_EXPRESSIONS => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_PARENTHESIZED_EXPRESSION => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_CALL => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_CALL_ARGUMENTS => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_ARGUMENT_LIST => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_EXPRESSION_LIST => RSyntaxKind::R_BOGUS,
            RSyntaxKind::__LAST => RSyntaxKind::R_BOGUS,
            RSyntaxKind::R_BOGUS => RSyntaxKind::R_BOGUS,
        }
    }

    #[inline]
    fn to_raw(&self) -> RawSyntaxKind {
        RawSyntaxKind(*self as u16)
    }

    #[inline]
    fn from_raw(raw: RawSyntaxKind) -> Self {
        Self::from(raw.0)
    }

    fn is_root(&self) -> bool {
        matches!(self, RSyntaxKind::R_ROOT)
    }

    fn is_list(&self) -> bool {
        RSyntaxKind::is_list(*self)
    }

    fn is_trivia(self) -> bool {
        matches!(
            self,
            RSyntaxKind::NEWLINE | RSyntaxKind::WHITESPACE | RSyntaxKind::COMMENT
        )
    }

    fn to_string(&self) -> Option<&'static str> {
        RSyntaxKind::to_string(self)
    }
}

impl TryFrom<RSyntaxKind> for TriviaPieceKind {
    type Error = ();

    fn try_from(value: RSyntaxKind) -> Result<Self, Self::Error> {
        if value.is_trivia() {
            match value {
                RSyntaxKind::NEWLINE => Ok(TriviaPieceKind::Newline),
                RSyntaxKind::WHITESPACE => Ok(TriviaPieceKind::Whitespace),
                RSyntaxKind::COMMENT => Ok(TriviaPieceKind::SingleLineComment),
                _ => unreachable!("Not Trivia"),
            }
        } else {
            Err(())
        }
    }
}

/// Text of `token`, excluding all trivia and removing quotes if `token` is a string literal.
pub fn inner_string_text(token: &RSyntaxToken) -> TokenText {
    let mut text = token.token_text_trimmed();
    if token.kind() == RSyntaxKind::R_STRING_VALUE {
        // remove string delimiters
        // SAFETY: string literal token have a delimiters at the start and the end of the string
        let range = TextRange::new(1.into(), text.len() - TextSize::from(1));
        text = text.slice(range);
    }
    text
}
