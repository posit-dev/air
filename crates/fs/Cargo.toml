
[package]
authors.workspace = true
categories.workspace = true
description = "Air filesystem utilities"
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
name = "fs"
publish = false
repository.workspace = true
rust-version.workspace = true
version = "0.0.0"

[dependencies]
path-absolutize = { workspace = true, features = [
    "once_cell_cache",
    "use_unix_paths_on_wasm",
] }

[lints]
workspace = true
