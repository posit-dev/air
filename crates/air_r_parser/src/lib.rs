mod error;
mod options;
mod parse;

#[allow(unused)]
mod treesitter;

use air_r_factory::RSyntaxFactory;
pub use error::ParseError;
pub use options::RParserOptions;
pub use parse::parse;
pub use parse::parse_r_with_cache;
pub use parse::Parse;

use air_r_syntax::RLanguage;
use biome_parser::tree_sink::LosslessTreeSink;

pub(crate) type RLosslessTreeSink<'source> = LosslessTreeSink<'source, RLanguage, RSyntaxFactory>;
