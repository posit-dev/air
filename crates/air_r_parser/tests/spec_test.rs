use air_r_parser::ParseError;
use air_r_parser::{parse, RParserOptions};
use air_r_syntax::{RRoot, RSyntaxNode};
use biome_console::fmt::{Formatter, Termcolor};
use biome_console::markup;
use biome_diagnostics::display::PrintDiagnostic;
use biome_diagnostics::termcolor;
use biome_diagnostics::DiagnosticExt;
use biome_parser::prelude::ParseDiagnostic;
use biome_rowan::SyntaxNode;
use biome_rowan::SyntaxSlot;
use biome_rowan::{AstNode, SyntaxKind};
use std::fmt::Write;
use std::fs;
use std::path::Path;

#[derive(Copy, Clone)]
pub enum ExpectedOutcome {
    Pass,
    Fail,
    Undefined,
}

#[ignore]
#[test]
fn quick_test() {
    let code = "
    1$foo
    1$2";
    let options = RParserOptions::default();
    let parsed = parse(code, options);
    let root = RRoot::unwrap_cast(parsed.syntax());
    println!("{root:#?}");
}

pub fn run(test_case: &str, _snapshot_name: &str, test_directory: &str, outcome_str: &str) {
    let outcome = match outcome_str {
        "ok" => ExpectedOutcome::Pass,
        "error" => ExpectedOutcome::Fail,
        "undefined" => ExpectedOutcome::Undefined,
        _ => panic!("Invalid expected outcome {outcome_str}"),
    };

    let test_case_path = Path::new(test_case);

    let file_name = test_case_path
        .file_name()
        .expect("Expected test to have a file name")
        .to_str()
        .expect("File name to be valid UTF8");

    let content = fs::read_to_string(test_case_path)
        .expect("Expected test path to be a readable file in UTF8 encoding");

    // Normalize to Unix line endings
    let content = line_ending::normalize(content);

    let options = RParserOptions::default();
    let parsed = parse(&content, options);
    let root = RRoot::unwrap_cast(parsed.syntax());
    let formatted_ast = format!("{:#?}", root);

    let mut snapshot = String::new();
    writeln!(snapshot, "\n## Input\n\n```R\n{content}\n```\n\n").unwrap();

    writeln!(
        snapshot,
        r#"## AST

```
{formatted_ast}
```

## CST

```
{:#?}
```
"#,
        parsed.syntax()
    )
    .unwrap();

    if parsed.has_errors() {
        let diagnostics: Vec<ParseDiagnostic> = parsed
            .errors()
            .iter()
            .map(ParseError::diagnostic)
            .map(ParseDiagnostic::clone)
            .collect();

        let mut diagnostics_buffer = termcolor::Buffer::no_color();

        let termcolor = &mut Termcolor(&mut diagnostics_buffer);
        let mut formatter = Formatter::new(termcolor);

        for diagnostic in diagnostics {
            let error = diagnostic
                .clone()
                .with_file_path(file_name)
                .with_file_source_code(&content);

            formatter
                .write_markup(markup! {
                    {PrintDiagnostic::verbose(&error)}
                })
                .expect("failed to emit diagnostic");
        }

        let formatted_diagnostics =
            std::str::from_utf8(diagnostics_buffer.as_slice()).expect("non utf8 in error buffer");

        if matches!(outcome, ExpectedOutcome::Pass) {
            panic!("Expected no errors to be present in a test case that is expected to pass but the following diagnostics are present:\n{formatted_diagnostics}")
        }

        writeln!(snapshot, "## Diagnostics\n\n```").unwrap();
        snapshot.write_str(formatted_diagnostics).unwrap();

        writeln!(snapshot, "```\n").unwrap();
    }

    match outcome {
        ExpectedOutcome::Pass => {
            let missing_required = formatted_ast.contains("missing (required)");
            if missing_required
                || parsed
                    .syntax()
                    .descendants()
                    .any(|node: RSyntaxNode| node.kind().is_bogus())
            {
                panic!("Parsed tree of a 'OK' test case should not contain any missing required children or bogus nodes: \n {formatted_ast:#?} \n\n {formatted_ast}");
            }

            let syntax = parsed.syntax();
            if has_bogus_nodes_or_empty_slots(&syntax) {
                panic!("modified tree has bogus nodes or empty slots:\n{syntax:#?} \n\n {syntax}")
            }
        }
        ExpectedOutcome::Fail => {
            if !parsed.has_errors() {
                panic!("Failing test must have diagnostics");
            }
        }
        _ => {}
    }

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => &test_directory,
    }, {
        insta::assert_snapshot!(file_name, snapshot);
    });
}

/// This check is used in the parser test to ensure it doesn't emit
/// bogus nodes without diagnostics.
pub fn has_bogus_nodes_or_empty_slots<L: biome_rowan::Language>(node: &SyntaxNode<L>) -> bool {
    node.descendants().any(|descendant| {
        let kind = descendant.kind();
        if kind.is_bogus() {
            return true;
        }

        if kind.is_list() {
            return descendant
                .slots()
                .any(|slot| matches!(slot, SyntaxSlot::Empty { .. }));
        }

        false
    })
}
