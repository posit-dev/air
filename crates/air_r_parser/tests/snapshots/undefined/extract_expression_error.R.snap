---
source: crates/air_r_parser/tests/spec_test.rs
expression: snapshot
---
## Input

```R
# TODO: tree-sitter-r currently allows an `optional()` RHS, but our grammar
# requires a RHS, so we get `missing (required)`. These should eventually be
# parse errors if we can switch tree-sitter-r away from using `optional()`.
# For example, `a$1` gives:

#> RExtractExpression {
#>     left: RDoubleValue {
#>         value_token: R_DOUBLE_LITERAL@10..16 "1" [Newline("\n"), Whitespace("    ")] [],
#>     },
#>     operator: DOLLAR@16..17 "$" [] [],
#>     right: missing (required),
#> },
#> RDoubleValue {
#>     value_token: R_DOUBLE_LITERAL@17..18 "2" [] [],
#> },

a$1
a$NA
a$NULL
a$TRUE
a$(b)
a${ b }

a@1
a@NA
a@NULL
a@TRUE
a@(b)
a@{ b }

```


## AST

```
RRoot {
    bom_token: missing (optional),
    expressions: RExpressionList [
        RExtractExpression {
            left: RIdentifier {
                name_token: IDENT@0..575 "a" [Comments("# TODO: tree-sitter-r ..."), Newline("\n"), Comments("# requires a RHS, so  ..."), Newline("\n"), Comments("# parse errors if we  ..."), Newline("\n"), Comments("# For example, `a$1`  ..."), Newline("\n"), Newline("\n"), Comments("#> RExtractExpression {"), Newline("\n"), Comments("#>     left: RDoubleV ..."), Newline("\n"), Comments("#>         value_toke ..."), Newline("\n"), Comments("#>     },"), Newline("\n"), Comments("#>     operator: DOLL ..."), Newline("\n"), Comments("#>     right: missing ..."), Newline("\n"), Comments("#> },"), Newline("\n"), Comments("#> RDoubleValue {"), Newline("\n"), Comments("#>     value_token: R ..."), Newline("\n"), Comments("#> },"), Newline("\n"), Newline("\n")] [],
            },
            operator: DOLLAR@575..576 "$" [] [],
            right: missing (required),
        },
        RDoubleValue {
            value_token: R_DOUBLE_LITERAL@576..577 "1" [] [],
        },
        RExtractExpression {
            left: RIdentifier {
                name_token: IDENT@577..579 "a" [Newline("\n")] [],
            },
            operator: DOLLAR@579..580 "$" [] [],
            right: missing (required),
        },
        RNaExpression {
            value: NA_LOGICAL_KW@580..582 "NA" [] [],
        },
        RExtractExpression {
            left: RIdentifier {
                name_token: IDENT@582..584 "a" [Newline("\n")] [],
            },
            operator: DOLLAR@584..585 "$" [] [],
            right: missing (required),
        },
        RNullExpression {
            null_token: NULL_KW@585..589 "NULL" [] [],
        },
        RExtractExpression {
            left: RIdentifier {
                name_token: IDENT@589..591 "a" [Newline("\n")] [],
            },
            operator: DOLLAR@591..592 "$" [] [],
            right: missing (required),
        },
        RTrueExpression {
            true_token: TRUE_KW@592..596 "TRUE" [] [],
        },
        RCall {
            function: RExtractExpression {
                left: RIdentifier {
                    name_token: IDENT@596..598 "a" [Newline("\n")] [],
                },
                operator: DOLLAR@598..599 "$" [] [],
                right: missing (required),
            },
            arguments: RCallArguments {
                l_paren_token: L_PAREN@599..600 "(" [] [],
                items: RArgumentList [
                    RArgument {
                        name_clause: missing (optional),
                        value: RIdentifier {
                            name_token: IDENT@600..601 "b" [] [],
                        },
                    },
                ],
                r_paren_token: R_PAREN@601..602 ")" [] [],
            },
        },
        RExtractExpression {
            left: RIdentifier {
                name_token: IDENT@602..604 "a" [Newline("\n")] [],
            },
            operator: DOLLAR@604..605 "$" [] [],
            right: missing (required),
        },
        RBracedExpressions {
            l_curly_token: L_CURLY@605..606 "{" [] [],
            expressions: RExpressionList [
                RIdentifier {
                    name_token: IDENT@606..608 "b" [Whitespace(" ")] [],
                },
            ],
            r_curly_token: R_CURLY@608..610 "}" [Whitespace(" ")] [],
        },
        RExtractExpression {
            left: RIdentifier {
                name_token: IDENT@610..613 "a" [Newline("\n"), Newline("\n")] [],
            },
            operator: AT@613..614 "@" [] [],
            right: missing (required),
        },
        RDoubleValue {
            value_token: R_DOUBLE_LITERAL@614..615 "1" [] [],
        },
        RExtractExpression {
            left: RIdentifier {
                name_token: IDENT@615..617 "a" [Newline("\n")] [],
            },
            operator: AT@617..618 "@" [] [],
            right: missing (required),
        },
        RNaExpression {
            value: NA_LOGICAL_KW@618..620 "NA" [] [],
        },
        RExtractExpression {
            left: RIdentifier {
                name_token: IDENT@620..622 "a" [Newline("\n")] [],
            },
            operator: AT@622..623 "@" [] [],
            right: missing (required),
        },
        RNullExpression {
            null_token: NULL_KW@623..627 "NULL" [] [],
        },
        RExtractExpression {
            left: RIdentifier {
                name_token: IDENT@627..629 "a" [Newline("\n")] [],
            },
            operator: AT@629..630 "@" [] [],
            right: missing (required),
        },
        RTrueExpression {
            true_token: TRUE_KW@630..634 "TRUE" [] [],
        },
        RCall {
            function: RExtractExpression {
                left: RIdentifier {
                    name_token: IDENT@634..636 "a" [Newline("\n")] [],
                },
                operator: AT@636..637 "@" [] [],
                right: missing (required),
            },
            arguments: RCallArguments {
                l_paren_token: L_PAREN@637..638 "(" [] [],
                items: RArgumentList [
                    RArgument {
                        name_clause: missing (optional),
                        value: RIdentifier {
                            name_token: IDENT@638..639 "b" [] [],
                        },
                    },
                ],
                r_paren_token: R_PAREN@639..640 ")" [] [],
            },
        },
        RExtractExpression {
            left: RIdentifier {
                name_token: IDENT@640..642 "a" [Newline("\n")] [],
            },
            operator: AT@642..643 "@" [] [],
            right: missing (required),
        },
        RBracedExpressions {
            l_curly_token: L_CURLY@643..644 "{" [] [],
            expressions: RExpressionList [
                RIdentifier {
                    name_token: IDENT@644..646 "b" [Whitespace(" ")] [],
                },
            ],
            r_curly_token: R_CURLY@646..648 "}" [Whitespace(" ")] [],
        },
    ],
    eof_token: EOF@648..649 "" [Newline("\n")] [],
}
```

## CST

```
0: R_ROOT@0..649
  0: (empty)
  1: R_EXPRESSION_LIST@0..648
    0: R_EXTRACT_EXPRESSION@0..576
      0: R_IDENTIFIER@0..575
        0: IDENT@0..575 "a" [Comments("# TODO: tree-sitter-r ..."), Newline("\n"), Comments("# requires a RHS, so  ..."), Newline("\n"), Comments("# parse errors if we  ..."), Newline("\n"), Comments("# For example, `a$1`  ..."), Newline("\n"), Newline("\n"), Comments("#> RExtractExpression {"), Newline("\n"), Comments("#>     left: RDoubleV ..."), Newline("\n"), Comments("#>         value_toke ..."), Newline("\n"), Comments("#>     },"), Newline("\n"), Comments("#>     operator: DOLL ..."), Newline("\n"), Comments("#>     right: missing ..."), Newline("\n"), Comments("#> },"), Newline("\n"), Comments("#> RDoubleValue {"), Newline("\n"), Comments("#>     value_token: R ..."), Newline("\n"), Comments("#> },"), Newline("\n"), Newline("\n")] []
      1: DOLLAR@575..576 "$" [] []
      2: (empty)
    1: R_DOUBLE_VALUE@576..577
      0: R_DOUBLE_LITERAL@576..577 "1" [] []
    2: R_EXTRACT_EXPRESSION@577..580
      0: R_IDENTIFIER@577..579
        0: IDENT@577..579 "a" [Newline("\n")] []
      1: DOLLAR@579..580 "$" [] []
      2: (empty)
    3: R_NA_EXPRESSION@580..582
      0: NA_LOGICAL_KW@580..582 "NA" [] []
    4: R_EXTRACT_EXPRESSION@582..585
      0: R_IDENTIFIER@582..584
        0: IDENT@582..584 "a" [Newline("\n")] []
      1: DOLLAR@584..585 "$" [] []
      2: (empty)
    5: R_NULL_EXPRESSION@585..589
      0: NULL_KW@585..589 "NULL" [] []
    6: R_EXTRACT_EXPRESSION@589..592
      0: R_IDENTIFIER@589..591
        0: IDENT@589..591 "a" [Newline("\n")] []
      1: DOLLAR@591..592 "$" [] []
      2: (empty)
    7: R_TRUE_EXPRESSION@592..596
      0: TRUE_KW@592..596 "TRUE" [] []
    8: R_CALL@596..602
      0: R_EXTRACT_EXPRESSION@596..599
        0: R_IDENTIFIER@596..598
          0: IDENT@596..598 "a" [Newline("\n")] []
        1: DOLLAR@598..599 "$" [] []
        2: (empty)
      1: R_CALL_ARGUMENTS@599..602
        0: L_PAREN@599..600 "(" [] []
        1: R_ARGUMENT_LIST@600..601
          0: R_ARGUMENT@600..601
            0: (empty)
            1: R_IDENTIFIER@600..601
              0: IDENT@600..601 "b" [] []
        2: R_PAREN@601..602 ")" [] []
    9: R_EXTRACT_EXPRESSION@602..605
      0: R_IDENTIFIER@602..604
        0: IDENT@602..604 "a" [Newline("\n")] []
      1: DOLLAR@604..605 "$" [] []
      2: (empty)
    10: R_BRACED_EXPRESSIONS@605..610
      0: L_CURLY@605..606 "{" [] []
      1: R_EXPRESSION_LIST@606..608
        0: R_IDENTIFIER@606..608
          0: IDENT@606..608 "b" [Whitespace(" ")] []
      2: R_CURLY@608..610 "}" [Whitespace(" ")] []
    11: R_EXTRACT_EXPRESSION@610..614
      0: R_IDENTIFIER@610..613
        0: IDENT@610..613 "a" [Newline("\n"), Newline("\n")] []
      1: AT@613..614 "@" [] []
      2: (empty)
    12: R_DOUBLE_VALUE@614..615
      0: R_DOUBLE_LITERAL@614..615 "1" [] []
    13: R_EXTRACT_EXPRESSION@615..618
      0: R_IDENTIFIER@615..617
        0: IDENT@615..617 "a" [Newline("\n")] []
      1: AT@617..618 "@" [] []
      2: (empty)
    14: R_NA_EXPRESSION@618..620
      0: NA_LOGICAL_KW@618..620 "NA" [] []
    15: R_EXTRACT_EXPRESSION@620..623
      0: R_IDENTIFIER@620..622
        0: IDENT@620..622 "a" [Newline("\n")] []
      1: AT@622..623 "@" [] []
      2: (empty)
    16: R_NULL_EXPRESSION@623..627
      0: NULL_KW@623..627 "NULL" [] []
    17: R_EXTRACT_EXPRESSION@627..630
      0: R_IDENTIFIER@627..629
        0: IDENT@627..629 "a" [Newline("\n")] []
      1: AT@629..630 "@" [] []
      2: (empty)
    18: R_TRUE_EXPRESSION@630..634
      0: TRUE_KW@630..634 "TRUE" [] []
    19: R_CALL@634..640
      0: R_EXTRACT_EXPRESSION@634..637
        0: R_IDENTIFIER@634..636
          0: IDENT@634..636 "a" [Newline("\n")] []
        1: AT@636..637 "@" [] []
        2: (empty)
      1: R_CALL_ARGUMENTS@637..640
        0: L_PAREN@637..638 "(" [] []
        1: R_ARGUMENT_LIST@638..639
          0: R_ARGUMENT@638..639
            0: (empty)
            1: R_IDENTIFIER@638..639
              0: IDENT@638..639 "b" [] []
        2: R_PAREN@639..640 ")" [] []
    20: R_EXTRACT_EXPRESSION@640..643
      0: R_IDENTIFIER@640..642
        0: IDENT@640..642 "a" [Newline("\n")] []
      1: AT@642..643 "@" [] []
      2: (empty)
    21: R_BRACED_EXPRESSIONS@643..648
      0: L_CURLY@643..644 "{" [] []
      1: R_EXPRESSION_LIST@644..646
        0: R_IDENTIFIER@644..646
          0: IDENT@644..646 "b" [Whitespace(" ")] []
      2: R_CURLY@646..648 "}" [Whitespace(" ")] []
  2: EOF@648..649 "" [Newline("\n")] []

```
